Fix formatting of linter messages

Hudda opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

__1. This PR fixes or fixes part of #[fill_in_number_here].__
2. This PR does the following: Fix the formatting of linter messages and remove the printing of messages as a list.

**Linter messages**
**Before**
![Screenshot from 2020-07-16 15-41-22](https://user-images.githubusercontent.com/29057268/87662341-ae768b80-c77f-11ea-970d-a38e0e525700.png)
![Screenshot from 2020-07-16 15-54-12](https://user-images.githubusercontent.com/29057268/87662551-feede900-c77f-11ea-83f3-4a23debe5ac2.png)

**After**
![Screenshot from 2020-07-16 15-27-19](https://user-images.githubusercontent.com/29057268/87662584-0a411480-c780-11ea-82c9-3992a1ae75fd.png)
![Screenshot from 2020-07-16 15-44-11](https://user-images.githubusercontent.com/29057268/87662606-12994f80-c780-11ea-98dd-e941e1cce830.png)

**Before**
![Screenshot from 2020-07-16 15-08-18](https://user-images.githubusercontent.com/29057268/87662960-af5bed00-c780-11ea-8c8a-c5c864ad7a0d.png)


**After**
![Screenshot from 2020-07-16 15-27-50](https://user-images.githubusercontent.com/29057268/87662895-93f0e200-c780-11ea-92b2-e6cdf91e3f03.png)



## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

Hudda committed:
Changed print to log

Hudda committed:
Changed run_backend_tests

Hudda commented:
@bansalnitish PTAL!

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -45,24 +45,19 @@ def log(message, show_time=False):
             self.output = self.func()
-            if self.verbose:
-                log('LOG %s:' % self.name, show_time=True)
-                log(self.output)
```
Why are you removing this? This is a general logger for concurrent tasks, see if you can fix the issue without changing this?

Hudda replied:
We don't need this now, because we are printing where we are running it, I tried to fix this by checking if we are getting a list of messages and ignore if we are getting a list but @seanlip disapproved that solution as that was not conceptually right(Sorry, I do not have the proof now, as the comment was resolved.)

DubeySandeep replied:
>  ignore if we are getting a list but @seanlip disapproved that solution as that was not conceptually right

Yeah, type checking is really not a good thing to do here!

I had a discussion with @bansalnitish on this today and we though this:

1. Have a function in linter class which will return a list of the linter functions.
2. In the pre_commit linter, file create a full list of these returned linter functions
3. Create a TaskThread object for each of these linter functions.
4. Run all the tasks in parallel.

(This way all the linter methods won't have to manage logging!)

Note: We have to find a way to run run "js_ts validator, parser" before running the lint test (not sure!)

You can discuss this with @bansalnitish in detail.

bansalnitish replied:
Hi @Hudda,

I think what Sandeep's said is fine (I know it would not be easy and would take some time to implement such stuff).

Let me explain the problem with the example. We will consider the `check _check_non_test_files` mentioned here: https://github.com/oppia/oppia/blob/develop/scripts/linters/python_linter.py#L108.

In my opinion, I see the following problem with this:

1. We have self.verbose and then the print statement saying this check is started. We would have to write the same three four lines while we write any other check. 
I would like to see this being removed from each of the check and added it somewhere in common (basically somehow use concurrent utils).

2. We see two `PRINT` statement in this check -- once while the error is caught and once during the end of the check. Do we really want these PRINT statements? In addition, we are returning the summary_messages which we will be printed at the last + message list would also be printed for checks in these functions while tasks are created via concurrent_utils. So, in total an error in this message would be seen four times. Do you see this as a problem?
I would like to have two messages being printed out of this test. Once while the entire list of messages is printed at the end by passing summary_messages (we won't touch that flow). Once, we could list the errors that this check caught at the end of this check. The check should have not have any PRINT statement.


This is how I think could be done (we can discuss in detail and this will require some design analysis):

1. We can remove all the PRINT statements from the check
2. Create a task thread of all functions. We can change this line: https://github.com/oppia/oppia/blob/develop/scripts/linters/python_linter.py#L599 to return all functions basically change it to

```
return custom_linter.get_linter_functions(), third_party_linter.get_linter_functions
```

3. Accumulate all these function in pre_commit_linter or wherever suitable. Execute them parallelly using concurrent utils, which we print the output of each test once it is completed (In addition, we can convert these private check functions to public and we will be able to execute all the checks in parallel rather than executing classes in parallel -- I might be wrong because I don't understand why these functions are marked private).

This would fix both the problems.

So, two questions for you:
1. Do you see removing all print statements which are scattered in each check, do not help much in debugging, addition of 4 lines at the start to check verbose condition etc is a problem (as mentioned in points 1-2 above) and should be solved (let us not consider how it would be implemented)?
2. Do you think we can achieve the approach that I mentioned?

Tagging @seanlip @DubeySandeep too for there thoughts!

Hudda replied:
I agree with the printing part, but running every method in parallel, I don't think this is a good idea to run each method in parallel as most of the methods depends on each other in class.

seanlip replied:
This was discussed in the most recent meeting with @bansalnitish and @Hudda.

Could one of you please add a clear, full summary of the plan of action and the results of the discussion, so that @DubeySandeep can take a look and comment?

bansalnitish replied:
OK, I can take this up.

As per our meeting:
1. We all agreed that the problems I posted above are genuine.
2. We decided to eradicate all the PRINT statements from the checks.
3. The checks would remain private and hence classes would be executed in parallel.
4. There would be no verbose statement too at the start of each check, it would be covered by concurrent_utils.
5. Each check would log two times, once while the checks run via concurrent_utils and once when the full list of summary_messages is printed at the end.
6. The API exposed by concurrent_utils should be such that it takes type from both the backend tests and the linter. Currently, the linter passed a list of string and backend tests pass strings. We will make this consistent so that both of them pass a similar object(it could be anything). 
7. If time allows (or may be in next milestone), we will even convert the summary_message that is returned from each check to a domain object with properties like check_name, status, errors, etc.

@Hudda could you please add if I missed something? 

Also, @DubeySandeep does this sounds good to you?

Hudda replied:
@DubeySandeep ping!

DubeySandeep replied:
> Also, @DubeySandeep does this sounds good to you?


This sounds awesome! Thanks for discussing this with others and finalizing it! :)

DubeySandeep replied:
> The checks would remain private and hence classes would be executed in parallel.

We had planned to run each linter function in parallel, right?

bansalnitish replied:
Yes, I agree. Though, this is more of design decision I would say. 

Anshul say's its fine to use classes in parallel. I'm fine with both to be honest, but I'm more inclined towards running functions in parallel. @Hudda can you please enlist why you don't want the functions to be converted to public?

Thanks!

DubeySandeep reviewed (changes requested):
[Marking this request changes!]

Hudda committed:
Fixed backend tests

bansalnitish commented:
@Hudda could you please add the logic behind this so that each reviewer has some context over this? Otherwise each reviewer like Sandeep would not have much context.

Write it along the way:
1. The backend tests use verbose mode or not (whatever you know better!).
2. The linter verbose mode just adds a print statement that this test is starting. 
3. Multiple messages were being printed after each class (add some code line links that cause this example: https://github.com/oppia/oppia/blob/develop/scripts/linters/pre_commit_linter.py#L541) when verbose mode was enabled because the function in concurrent_utils does [this].
4. Given this, we faced [this] problem.
5. Now, it removed the verbose mode from the concurrent_utils which will affect this[mention the affect].
6. To counter this, I made [this] changes in the linter and I made [this] change in the backend test file. It will fix our problem and not alter the backend tests output because of [this].

(Currently, it is just me and Sean who know that we decided we won't be adding any if condition in the concurrent_utils to avoid printing of message list)

It would take not more than 10 -15 minutes I guess to fill the part in `[]` above. Please ping me over chat when this is done to take a full pass or in case you need any help. Thanks!

Hudda commented:
1. Backend tests use verbose to show the name of tests it is currently running and the status of tests.
2. Linter verbose mode just adds a print statement with the name of the test.
3. concurrent_task_utils is printing the result we get from the linter or run_backend_test when the verbose mode is enabled. (https://github.com/oppia/oppia/blob/f3ec0b4e15059ab8e6b452abca8bb9e59b6292bf/scripts/concurrent_task_utils.py#L62)
4. Because of this, we are printing a list of messages after every linter class finish linting.
5. Now, I removed the verbose mode from the concurrent_utils_task due to which we are no longer printing a list of messages after every class.
6. But this is affecting the working of backend tests because now backend tests no longer have verbose mode, to fix this problem, I've added the verbose mode in the run_backend_tests module itself.
7. Also, we had a problem that linter is printing multiple messages on a single line, to fix this I used log method of concurrent_tasks_util, because it is written for the purpose of replacing  `print` method of python in case of multiprocessing. Now, every message is printing on a separate line

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

DubeySandeep commented:
@bansalnitish, Can you please review this PR first as I'm not sure about the expectations of this PR?

bansalnitish commented:
I don't think it is ready for review right now!

Hudda commented:
Design is not finalized yet, @DubeySandeep has some comments

seanlip commented:
@bansalnitish @Hudda we talked about this in the meeting. Could one of you please explain clearly the remaining list of TODOs for this PR? (Maybe as a checklist?)

bansalnitish commented:
Anshul will implement these 7 points: https://github.com/oppia/oppia/pull/9925#discussion_r458912925 + will make the checks public.

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Temp

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Changed css linter structure

Hudda committed:
Changed structure

Hudda committed:
temp

bansalnitish commented:
@Hudda is this PR ready for review?

Hudda commented:
> @Hudda is this PR ready for review?

No, this is not ready yet, some of the files still needed to changed.

bansalnitish commented:
Hi @Hudda, could you please update me on this -- how is it going? Do you need my help?

Hudda commented:
> Hi @Hudda, could you please update me on this -- how is it going? Do you need my help?

Thanks, but still working on #10107

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Changed structure of sublinters

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
temp

bansalnitish commented on scripts/linters/python_linter.py:
```diff
@@ -127,10 +125,10 @@ def _check_non_test_files(self):
                         summary_messages.append(summary_message)
                         failed = True
 
+            full_messages = summary_messages
```
full_messages is not initialised.

Hudda replied:
Initialised

Hudda committed:
Inital changes

ankita240796 commented:
@Hudda please ping me here once you want me to take a pass. Thanks!

Hudda committed:
temp

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Changed structure of linter

Hudda committed:
Changed tests

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda commented:
@bansalnitish PTAL!

Hudda committed:
Added lint tests

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Addressed review comments

Hudda committed:
Fix backend tests

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

bansalnitish commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,6 +44,45 @@ def log(message, show_time=False):
+        """Constructs a OutputStream object.
+
+        Args:
+            name: str. Name of the check currently running.
```
Since the output stream would be returned after the check runs -- why have we written this as "check currently running"?

Hudda replied:
removed currently.

bansalnitish commented on scripts/linters/codeowner_linter.py:
```diff
@@ -14,19 +14,17 @@
 # limitations under the License.
 
-"""Codeowner file checks."""
+"""Lint checks for codeowner file."""
```
file -> files

Hudda replied:
It should be file instead of files because we have a single file only

bansalnitish commented on scripts/linters/js_ts_linter.py:
```diff
@@ -114,33 +113,22 @@ def compile_all_ts_files():
 class JsTsLintChecksManager(python_utils.OBJECT):
-    """Manages all the Js and Ts linting functions.
-
-    Attributes:
```
Why is this comment removed?

Hudda replied:
It was duplicate we have all these listed in __init__() method.

bansalnitish commented on scripts/linters/js_ts_linter.py:
```diff
@@ -171,12 +159,8 @@ def _validate_and_parse_js_and_ts_files(self):
         parsed_js_and_ts_files = dict()
-        if not self.verbose_mode_enabled:
-            python_utils.PRINT('Validating and parsing JS and TS files ...')
+        concurrent_task_utils.log('Validating and parsing JS and TS files ...')
```
This line should be printed when the verbose mode is not enabled in the original code. However, you're printing it each time this check runs. Any reason?

Hudda replied:
I think it will be useful to know that we are parsing the files in both cases before we are printing the name of each file in verbose mode since we are not doing that anymore so it will be helpful to print this in both cases.

bansalnitish replied:
So, this statement should be there in each check right?  I mean all the checks should be consistent.

Hudda replied:
No, This method will run just once at the beginning of JsTs linter and this is not a check, this is for parsing all files. So, I don't think we need to be consistent here.

bansalnitish commented on scripts/linters/js_ts_linter.py:
```diff
@@ -238,570 +222,459 @@ def _get_compiled_ts_filepath(self, filepath):
+            OutputStream. An OutputStream object to retrieve the status of a
+            lint check.
+        """
+        name = 'Ts ignore'
```
Can we please follow some convention for naming the tests too? 

One way would be converting this camelcase `_check_ts_ignore` to `TS Ignore` -- start with capital when it has dash before it.

Hudda replied:
I tried doing this and it looks very weird to have letter in caps in the middle. Example: `SUCCESS Ts Ignore checks passed`

bansalnitish replied:
Oh yes, you're right thanks!

bansalnitish commented on scripts/linters/js_ts_linter.py:
```diff
@@ -238,570 +222,459 @@ def _get_compiled_ts_filepath(self, filepath):
+        return concurrent_task_utils.OutputStream(
+            name, failed, summary_messages, summary_messages)
 
     def _check_ts_expect_error(self):
```
Why are these checks not made public unlike other linter files?

Hudda replied:
Because in this linter we have to parse and compile all files before using them in any of the checks, in other words, the checks are dependent, we make these public and test these independently then this will create the problems. Plus, we will migrate all of these to custom eslint checks, therefore I am not seeing any benefit in making these checks public.

bansalnitish commented on scripts/linters/other_linter.py:
```diff
@@ -0,0 +1,276 @@
+# coding: utf-8
```
The name of this files is not good. Can you please change this? 

Suggestion: Can we keep the original files instead of merging these three checks together and naming it other_linter? We can have a class in each of those original files similar to what you did for codeowner linter.

Hudda replied:
Changed the name to other_file_linter, I did this because all the linter files are named by the file types, for example, python_linter, css_linter, etc. And we have other file types for all the remaining file types, so I think it is better to have this name

bansalnitish commented on scripts/linters/other_linter_test.py:
```diff
@@ -0,0 +1,246 @@
+# coding: utf-8
```
Ditto for this file too.

Hudda replied:
Changed to other_file _linter_test.py

bansalnitish reviewed (commented):
Took a first pass -- this is a big PR and might require couple of more passes. I have not reviewed any test file right now. Please fix the failing tests and ping me back!

On the other hand, I'm very happy to see these changes. We do not have stdout/ sys.stdout/ streaming objects now. We do not have print statements messed up all around. A lot of code has been removed which has seriously made these files look a lot better. Thanks @Hudda -- I really appreciate this work!

Tagging @DubeySandeep too -- to may be take a pass at any one linter.py file.

Hudda committed:
Addressed review comments

Hudda committed:
Changed file name

Hudda commented:
@bansalnitish PTAL!

Hudda committed:
Fixed lint errors

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,6 +44,45 @@ def log(message, show_time=False):
             python_utils.PRINT(message)
 
 
+class OutputStream(python_utils.OBJECT):
```
Considering this represents result/output of the task so should we call this TaskOutput or TaskResult (The later one sounds more like a class name.)?

Hudda replied:
Changed to TaskResult.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,6 +44,45 @@ def log(message, show_time=False):
+        self.full_messages = full_messages
+
+    @property
+    def all_messages(self):
```
I'm confused with the properties `messages`, `full_messages` & `all_messages`, can you please add a better name? (Let me know what each represents I can help you find a better name)

Hudda replied:
`messages` contains the messages after stripped by the stripper functions.
`full_messages` conatin unstripped messages.
`all_messages` all pass or fail status.

bansalnitish replied:
@DubeySandeep ping!

DubeySandeep replied:
`all_messages all pass or fail status.`

Sorry, I didn't get this can you please rephrase it?

Also, can you explain how these messages are generated and how it's used? [Asking this as I'm unable to understand these things with the variable name!]

Hudda replied:
`messages` contain all the stripped error messages, these messages are those which we get after removing the extra part from the error messages. We are going to use these messages for non-verbose mode.
`full_messages` contain messages without any stripped part. These messages are going to be used for verbose mode.
`all_messages` here, we add pass or fail status to the full_messages. these are also used to verbose mode.

DubeySandeep replied:
Change the variable name similarly: Should we have this: `full_messages` & `trimmed_messages`?

Hudda replied:
Done.

DubeySandeep replied:
I don't see any changes, I think you haven't changed the `self.message` to `self.trimmed_message`

Hudda replied:
Changed now.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -60,12 +101,23 @@ def run(self):
-                log('----------------------------------------')
+                for stdout in self.output:
+                    if stdout.name:
+                        log('Starting %s check' % stdout.name, show_time=True)
```
This is not true, right? by this stage the checks have already run, right? If yes then either remove this or change it to `Completed running <name> check`

Hudda replied:
I think it is better to keep as it is because it gives the user a indication that the check is running as the output is going to be printed under this statement. And if we do what you suggested then it will show `Completed running <name> check` and then `Finished <check name>`. I think this is not a good idea.

bansalnitish replied:
I don't agree with you on this. A check that is finished cannot print saying `Starting check`.

bansalnitish replied:
We can change it to something "Printing errors from XYZ check".

Hudda replied:
Changed.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -60,12 +101,23 @@ def run(self):
+                        log('Starting %s check' % stdout.name, show_time=True)
+                        log('----------------------------------------')
+                        for message in stdout.all_messages:
+                            log(message)
```
On the terminal, there can be a log from another task in between these for loop, right? If yes, then please add a `get_task_report` (or anything similar) in the TaskResult class which will return the big string which you can print in one call. Do you think this would be fine?

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -60,12 +101,23 @@ def run(self):
+                            show_time=True)
+                    else:
+                        log('LOG %s:' % self.name, show_time=True)
+                        log(stdout.all_messages)
```
For which case, we are doing this?

Hudda replied:
This will print backend tests result.

bansalnitish replied:
Can we please add a comment telling the first section is for lint and the other is for backend?

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/general_purpose_linter.py:
```diff
@@ -697,52 +686,30 @@ def _check_for_mandatory_pattern_in_file(
-            python_utils.PRINT('----------------------------------------')
-
+        name = 'Mandatory pattern'
         summary_messages = []
```
is this only going to store the error messages? if so then call it error_messages* [Ditto elsewhere!]

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/general_purpose_linter.py:
```diff
@@ -697,52 +686,30 @@ def _check_for_mandatory_pattern_in_file(
+                        pattern_list, filepath, failed))
+                summary_messages.extend(mandatory_summary_messages)
+        return concurrent_task_utils.OutputStream(
+            name, failed, summary_messages, summary_messages)
```
Why repeated values `summary_messages`?

Hudda replied:
the second last summary message is used to pass stripped messages in case if the current linter has a stripper method to strip the messages. but in this linter we do not have such method so we are passing the summary_message twice.

bansalnitish replied:
Can we use the same terminology to pass the messages? Like create full_messages in this check too and assign it equal to summary_messages and pass that? 

I would love to hear other suggestions too. Though, seeing two summary_messages would be really weird for anyone.

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/general_purpose_linter.py:
```diff
@@ -623,22 +620,15 @@ def check_file_type_specific_bad_pattern(filepath, content):
     """Manages all the common linting functions. As an abstract base class, this
     is not intended to be used directly.
-
-    Attributes:
```
Why removing these?

Hudda replied:
These are duplicates we have these in `__init__` method.

DubeySandeep reviewed (changes requested):
@Hudda, I took a quick pass on 2 files (util and genral_purpose_linter file) and left a few comments PTAL! 

In general, the pattern looks great we are going in a good direction! :) Make sure to create issues for all the works we are expecting to do in the future (As per our final design decisions). Thanks for the great work! :)

DubeySandeep commented:
@Hudda, Note faling backend covergae issue!

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Addressed review comments

Hudda commented:
@DubeySandeep PTAL!

DubeySandeep commented:
@bansalnitish, Can you please review this PR as a main review going ahead? I'm happy to take a final pass or provide any suggestions required [You can tag me in comments.]. Thanks! :)

Hudda commented:
@bansalnitish ping!

bansalnitish commented:
@Hudda I see there were not follow ups on some of the comments above. I followed at each one of them and left a comment on the required ones. Can you please take a pass over all these and ping me to review it fully back?

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Addressed review comments

Hudda committed:
Fixed backend tests

Hudda committed:
Addressed review comment

Hudda commented:
@bansalnitish PTAL!

DubeySandeep commented:
@Hudda, Assign the PR to @bansalnitish If you want him to review the PR!

bansalnitish commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,6 +44,54 @@ def log(message, show_time=False):
+        failing or passing.
+
+        Returns:
+            list(str). List of full messages corresponding to currently running
```
Currently running task? The task would have been finished right while we print these results?

Hudda replied:
Changed.

bansalnitish commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,6 +44,54 @@ def log(message, show_time=False):
+        Args:
+            name: str. Name of the check running.
+            failed: bool. Status of the check currently running.
+            error_messages: list(str). List of summary messages returned by
```
How would anyone know what is summary messages here? Please make it clear.

Something like:  `List of error messages that are stripped to keep XYZ in the message`.

Hudda replied:
Done.

bansalnitish commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,6 +44,54 @@ def log(message, show_time=False):
+
+    @property
+    def all_messages(self):
+        """Returns the full list with the message if the current check is
```
Returns a list of message for the current check

Hudda replied:
Done.

bansalnitish commented on scripts/linters/css_linter.py:
```diff
@@ -105,20 +95,16 @@ def _lint_css_files(self):
         num_files_with_errors = 0
-        summary_messages = []
+        failed = False
+        error_messages = []
```
Could we name error_messages to may be `stripped_error_messages`? Then it would be easy to differentiate between full vs stripped. Though, there could be some other better name.

Hudda replied:
Done.

bansalnitish commented on scripts/linters/other_files_linter.py:
```diff
@@ -0,0 +1,279 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Lint checks for codeowner file."""
```
Is this lint check for codeowner file only? I can see the codeowner_linter.py. 

Though I see you have deleted the app_dev_linter and have added checks from that file to this one.

Hudda replied:
Sorry, the file docstring was wrong changed now.

bansalnitish reviewed (commented):
Took another pass!

Hudda committed:
Addressed review comment

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Fixed docstring.

Hudda commented:
@bansalnitish PTAL!

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,6 +44,53 @@ def log(message, show_time=False):
+        if self.name:
+            failed_message = (
+                '%s %s check %s' % (
+                    (linter_utils.FAILED_MESSAGE_PREFIX, self.name, 'failed')
```
Are we expecting to use `FAILED_MESSAGE_PREFIX` to use anywhere in the linter file, if no then move this constant in this file!

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,6 +44,53 @@ def log(message, show_time=False):
+        return self.full_messages
+
+    @property
+    def get_task_report(self):
```
property name should not start with get*

Hudda replied:
removed get.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,6 +44,53 @@ def log(message, show_time=False):
 
 
+class TaskResult(python_utils.OBJECT):
+    """Output stream for concurrent_task_utils."""
```
Change the docstring of the class as per the class name*

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,6 +44,53 @@ def log(message, show_time=False):
+
+        Args:
+            name: str. Name of the check running.
+            failed: bool. Status of the check currently running.
```
The object from this class are created after the test completes, right? So let's change this to `The boolean value representing whether the task failed.`

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -60,12 +109,24 @@ def run(self):
         try:
             self.output = self.func()
```
self.task_results*

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -60,12 +109,24 @@ def run(self):
-                log('LOG %s:' % self.name, show_time=True)
-                log(self.output)
-                log('----------------------------------------')
+                for stdout in self.output:
```
`                for task_result in self.task_results:`

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -60,12 +109,24 @@ def run(self):
+                    # The following section will print the output of the lint
+                    # checks.
+                    if stdout.name:
+                        log(
```
Don't use `log` thrice, create a variable message_to_log and append data in that. In the end, use log(message_to_log)

Hudda replied:
I don't get this comment. In some messages we are logging time as well, If we use just a single log, then we have to either use show_time in all or don't use at all. And if we try to use show_time in all, then it won't look good with a line containing dashes(----) only.

DubeySandeep replied:
I'm talking about the next three lines, if you log them as one message with show_time=True it will be same as before, right? + it will ensure us that there won't be any other message in between, right?

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -60,12 +109,24 @@ def run(self):
+                    # checks.
+                    if stdout.name:
+                        log(
+                            'Printing %s check\'s status' % stdout.name,
```
May be `Report from <check-name> check`?

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -60,12 +109,24 @@ def run(self):
+                    # tests.
+                    else:
+                        log('LOG %s:' % self.name, show_time=True)
+                        log(stdout.all_messages)
```
Ditto as above use `log` once*

Hudda replied:
I don't get this comment. In some messages we are logging time as well, If we use just a single log, then we have to either use show_time in all or don't use at all. And if we try to use show_time in all, then it won't look good with a line containing dashes(----) only.

DubeySandeep reviewed (commented):
[Leaving part of review comments as per offline discussion.]

DubeySandeep commented on scripts/concurrent_task_utils_test.py:
```diff
@@ -67,7 +99,7 @@ class TaskThreadTests(ConcurrentTaskUtilsTests):
     def test_task_thread_with_success(self):
         task = concurrent_task_utils.TaskThread(
-            test_function('unused_arg'), True, self.semaphore, name='test')
+            test_function('unused_arg'), False, self.semaphore, name='test')
```
Why this change? I would suggest you have a separate test for verbose mode disabled?

Hudda replied:
I've already added tests with verbose mode enabled at line 123.

DubeySandeep commented on scripts/concurrent_task_utils_test.py:
```diff
@@ -79,20 +111,64 @@ def test_task_thread_with_success(self):
             '(0 given)' in self.linter_stdout)
 
+    def test_task_thread_with_verbose_mode_enabled(self):
+        class HelperTests(python_utils.OBJECT):
```
Add this to setUp method to avoid duplicates*

Hudda replied:
Both HelperTests classes have different codes, so can't add it in setUp method.

DubeySandeep replied:
optional: Have an init function to configure the object as needed.

DubeySandeep commented on scripts/concurrent_task_utils_test.py:
```diff
@@ -79,20 +111,64 @@ def test_task_thread_with_success(self):
+        with self.print_swap:
+            task.start()
+            task.join()
+        expected_output = [s for s in self.linter_stdout if 'Printing' in s]
```
Can we assert the whole output? (If date time is the issue, can we mock the log function and try?)

Hudda replied:
We can do that, but we also have dashes in the output. So I don't that it is a good idea to match the whole output

DubeySandeep commented on scripts/linters/codeowner_linter.py:
```diff
@@ -51,173 +49,170 @@
+            error_messages.append(error_message)
+            failed = True
+
+        return failed, error_messages
```
As these functions are now in the same class, I would suggest having self.failed and self.error_messages (instead of returning these values)

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/codeowner_linter.py:
```diff
@@ -288,39 +282,53 @@ def check_codeowner_file(file_cache, verbose_mode_enabled):
+
+
+def get_linters(file_cache):
+    """Creates ThirdPartyCSSLintChecksManager and returns it.
```
Update the docstring correctly*

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/codeowner_linter_test.py:
```diff
@@ -89,28 +89,30 @@ def test_missing_important_codeowner_path_from_list(self):
-                codeowner_linter.check_codeowner_file(FILE_CACHE, True)
+                linter = codeowner_linter.CodeOwnerLintChecksManager(FILE_CACHE)
+                linter_stdout = linter.check_codeowner_file()
         self.assert_same_list_elements([
```
Are we expecting to test the full log now?

Hudda replied:
We still have filename at the beginning of the message, which differs on every machine. So not matching the full path for now.

DubeySandeep replied:
Please assert the full log! You can use assertRegex if required!

DubeySandeep commented on scripts/linters/codeowner_linter.py:
```diff
@@ -51,173 +49,170 @@
-def _check_for_important_patterns_at_bottom_of_codeowners(important_patterns):
-    """Checks that the most important patterns are at the bottom
-    of the CODEOWNERS file.
+    def _check_for_important_patterns_at_bottom_of_codeowners(
```
are we still keeping these functions private?

Hudda replied:
I don't think there is a need to make code owner files function public as these are connected to each other, although I made functions public in all other files.

DubeySandeep commented on scripts/linters/codeowner_linter_test.py:
```diff
@@ -89,28 +89,30 @@ def test_missing_important_codeowner_path_from_list(self):
-                codeowner_linter.check_codeowner_file(FILE_CACHE, True)
+                linter = codeowner_linter.CodeOwnerLintChecksManager(FILE_CACHE)
+                linter_stdout = linter.check_codeowner_file()
         self.assert_same_list_elements([
```
Should we remove this function form the LinterTest class?

Hudda replied:
We still need it. So, can't remove it for now. But may remove after I migrate all custom js/ts linter checks to eslint checks.

DubeySandeep replied:
Please remove this function we had added this for temporary and thought of removing it completely in this PR!

bansalnitish replied:
Hi @DubeySandeep, I don't think this was supposed to be done in this PR. It will be done in separate PR where we will match the complete logs. Refer: https://github.com/oppia/oppia/pull/9925#discussion_r458912925

DubeySandeep replied:
We are changing the private linter functions to the public so that we can also write tests for each function and assert the complete error log reported by the function, that was the expectation, right? Are we thinking of doing this in a separate PR? I'm fine with that but I think the main expectations from these public functions are that we can run them separately and can have test complete logs and remove `assert_same_list_elements` function from the codebase completely!

bansalnitish replied:
Yes, the expectation is totally correct.

Yup, we plan to do it in a separate PR.

DubeySandeep replied:
> Yup, we plan to do it in a separate PR.

I'm fine with that, but please make sure to remove the `assert_same_list_elements` from the codebase before we start using it in other places! [It would be great if you can prioritize that PR, I think that was the major expectation from this PR.] 

EDIT:

That was = removing `assert_same_list_elements` and asserting the full log (This is the major issue in our linter test).

bansalnitish replied:
Sure.

DubeySandeep commented on scripts/linters/codeowner_linter_test.py:
```diff
@@ -89,28 +89,30 @@ def test_missing_important_codeowner_path_from_list(self):
-            codeowner_linter.check_codeowner_file(FILE_CACHE, True)
+        with self.listdir_swap, codeowner_path_swap:
+            linter = codeowner_linter.CodeOwnerLintChecksManager(FILE_CACHE)
+            linter_stdout = linter.check_codeowner_file()
```
This will return a TaskReport  object? If yes then rename the variables to lint_task_report and validate all the parts of the taskReport object*

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/codeowner_linter_test.py:
```diff
@@ -157,55 +161,59 @@ def test_check_codeowner_file_with_success_message(self):
-        self.assert_failed_messages_count(self.linter_stdout, 0)
+                linter = codeowner_linter.CodeOwnerLintChecksManager(FILE_CACHE)
+                linter_stdout = linter.check_codeowner_file()
+        self.assertEqual(
```
This assertion looks great! :)

Hudda committed:
Addressed review comments

DubeySandeep reviewed (commented):
[Another partial review comments]

Hudda committed:
Addressed review comments

Hudda commented:
@DubeySandeep PTAL!

DubeySandeep commented:
Assigning this PR to @bansalnitish as per our offline discussion!

bansalnitish commented on scripts/linters/css_linter.py:
```diff
@@ -137,50 +123,43 @@ def _lint_css_files(self):
+            failed = True
+
+        return concurrent_task_utils.TaskResult(
+            name, failed, stripped_error_messages, full_error_messages)
```
Can we please replicate the same comments of the `codeowner_linter.py` to everywhere else? (Make this self.failed etc.)

Hudda replied:
I've already replicated the codeowner_linter.py changes everywhere except self.failed and self.messages as in codeowner linter we have checks related to each other, while in other linters all checks are independent so we can't do that here.

bansalnitish commented on scripts/linters/css_linter.py:
```diff
@@ -88,11 +77,12 @@ def _get_trimmed_error_output(css_lint_output):
 
         Returns:
-            summary_messages: list(str). Return summary of lint checks.
+            TaskResult. An TaskResult object to retrieve the status of a
```
An -> A

Hudda replied:
Done.

bansalnitish commented on scripts/linters/js_ts_linter.py:
```diff
@@ -238,804 +222,662 @@ def _get_compiled_ts_filepath(self, filepath):
 
-        summary_messages = []
+        Returns:
+            TaskResult. An TaskResult object to retrieve the status of a
```
Ditto here and elsewhere.

Hudda replied:
Done.

bansalnitish reviewed (approved):
LGTM -- thanks @Hudda. The linter looks so much better now.

Hudda committed:
Addressed review comments

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda commented:
@DubeySandeep PTAL!

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,13 +44,60 @@ def log(message, show_time=False):
+        """Constructs a TaskResult object.
+
+        Args:
+            name: str. Name of the check running.
```
The name of the task. (This is the task report so the task must be completed till this object is created!)

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,13 +44,60 @@ def log(message, show_time=False):
+                    if self.failed else (
+                        SUCCESS_MESSAGE_PREFIX, self.name, 'passed')))
+            self.full_messages.append(failed_message)
+        return self.full_messages
```
Is it expected to change the full_message property here? IF yes, then why do we have these two property `full_message` and all_message?

Hudda replied:
We have to add the pass or fail status to the the full_message list, for that we need all_messages property

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,13 +44,60 @@ def log(message, show_time=False):
+        return self.full_messages
+
+    @property
+    def task_report(self):
```
Let's not have the name of the class and it's property be kinda same. I would suggest doing the join inline instead of having a separate property for it.

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -58,14 +107,28 @@ def __init__(self, func, verbose, semaphore, name):
+                        log(
+                            'LOG %s:\n%s\n'
+                            '----------------------------------------' %
+                            (self.name, task_result.all_messages),
```
`all_messages` is a list, right? is it expected to print a list here?

Hudda replied:
No, backend tests do not return a list, but a string.

DubeySandeep replied:
Make it return a list as suggest above in a comment*

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/codeowner_linter.py:
```diff
@@ -51,174 +49,165 @@
-    with linter_utils.redirect_stdout(stdout):
-        failed = False
-        summary_messages = []
+class CodeOwnerLintChecksManager(python_utils.OBJECT):
```
CodeownerLint...*

Hudda replied:
Done

DubeySandeep commented on scripts/linters/codeowner_linter.py:
```diff
@@ -51,174 +49,165 @@
+    """Manages codeowner checks."""
+
+    def __init__(self, file_cache):
+        """Constructs a PythonLintChecksManager object.
```
```suggestion
        """Constructs a CodeOwnerLintChecksManager object.
```

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils_test.py:
```diff
@@ -79,20 +111,64 @@ def test_task_thread_with_success(self):
+        with self.print_swap:
+            task.start()
+            task.join()
+        expected_output = [s for s in self.linter_stdout if 'Report' in s]
```
Assert the full output* (You can use assertRegex if required!) Also, try avoiding logics (if-else) in tests!

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils_test.py:
```diff
@@ -79,20 +111,64 @@ def test_task_thread_with_success(self):
+        with self.print_swap:
+            task.start()
+            task.join()
+        expected_output = [s for s in self.linter_stdout if 'FINISHED' in s]
```
Ditto*

Hudda replied:
The output is too big here, so it is not possible to assert full message.

DubeySandeep commented on scripts/linters/codeowner_linter.py:
```diff
@@ -51,174 +49,165 @@
+        bottom of the CODEOWNERS file.
+
+        Returns:
+            TaskResult. A TaskResult object to retrieve the status of a
```
```suggestion
            TaskResult. A TaskResult object representing the reslut of the lint check.
```

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/codeowner_linter.py:
```diff
@@ -288,39 +276,50 @@ def check_codeowner_file(file_cache, verbose_mode_enabled):
+
+        full_error_messages = self.error_messages
+        return concurrent_task_utils.TaskResult(
+            name, self.failed, self.error_messages, full_error_messages)
```
why you have created a variable on line 296? is that of any use?

Hudda replied:
NO, removed.

DubeySandeep commented on scripts/linters/codeowner_linter.py:
```diff
@@ -288,39 +276,50 @@ def check_codeowner_file(file_cache, verbose_mode_enabled):
+
+        Returns:
+            list(TaskResult). A list of TaskResult objects to be used for
+            linter status retrieval.
```
```suggestion
            list(TaskResult). A list of TaskResult objects representing the results of the lint checks.
```

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/codeowner_linter_test.py:
```diff
@@ -60,7 +60,7 @@
 
 class CodeOwnerLinterTests(test_utils.LinterTestBase):
-    """Test the methods for codeowner linter script."""
+    """Test the methods for Codeowner."""
```
```suggestion
    """TUnit test for the CodeOwnerLintChecksManager class."""
```

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/codeowner_linter_test.py:
```diff
@@ -89,28 +89,30 @@ def test_missing_important_codeowner_path_from_list(self):
-            '\' section.'], self.linter_stdout)
-        self.assert_failed_messages_count(self.linter_stdout, 1)
+            '\' section.'], lint_task_report.messages)
+        self.assert_failed_messages_count(lint_task_report.all_messages, 1)
```
assert lint_task_report.failed is True. (Also, assert the other property of lint_task_report)

Ditto elsewhere!

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/css_linter.py:
```diff
@@ -137,50 +123,43 @@ def _lint_css_files(self):
-            python_utils.PRINT('')
-            python_utils.PRINT(
+            concurrent_task_utils.log('')
+            concurrent_task_utils.log(
```
Let's try avoiding using log functions in these files. Can we return a taskResult object with the same message:
`return [concurrent_task_utils.TaskResult('CSS lint', False, 'There are no HTML or CSS files to lint.', '')]`

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/css_linter_test.py:
```diff
@@ -93,35 +92,29 @@ def mock_communicate():
-            self.linter_stdout)
-        self.assert_failed_messages_count(self.linter_stdout, 0)
+            CONFIG_PATH, [VALID_CSS_FILEPATH])
+        lint_task_report = third_party_linter.lint_css_files()
```
Are we expecting to call `perform_all_lint_checks` function here?

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/css_linter_test.py:
```diff
@@ -75,7 +74,7 @@ def mock_join(*unused_args):
         third_party_linter = css_linter.ThirdPartyCSSLintChecksManager(
-            CONFIG_PATH, [INVALID_CSS_FILEPATH], False)
+            CONFIG_PATH, [INVALID_CSS_FILEPATH])
         with self.print_swap, join_swap, self.assertRaisesRegexp(
```
Remove `print_swap` from all the places where it's not needed! (Do the clean-up thoroughly as these will create confusions later.)

Hudda replied:
We need it here, in case if there is an error while running the linter, we are printing the error and exiting the css linter. so to catch that message we need print_swap here.

DubeySandeep commented on scripts/linters/general_purpose_linter.py:
```diff
@@ -581,16 +580,14 @@ def check_bad_pattern_in_file(filepath, file_content, pattern):
-            return failed, summary_messages
-    return failed, summary_messages
+            return failed, error_messages
+    return failed, error_messages
```
Should these functions return taskResult? This will help us to avoid another big rework when lint function will run as a taskThread (Currently we run the class)

Hudda replied:
No, this is a helper function of the main function we do not need to return taskresult here.

DubeySandeep commented on scripts/linters/general_purpose_linter.py:
```diff
@@ -694,203 +683,142 @@ def _check_for_mandatory_pattern_in_file(
+                    self._check_for_mandatory_pattern_in_file(
+                        pattern_list, filepath, failed))
+                error_messages.extend(mandatory_error_messages)
+        full_error_messages = error_messages
```
is this assignment necessary?

Hudda replied:
Removed.

DubeySandeep commented on scripts/linters/general_purpose_linter.py:
```diff
@@ -694,203 +683,142 @@ def _check_for_mandatory_pattern_in_file(
-        return all_messages
+        if not self.all_filepaths:
+            concurrent_task_utils.log('There are no files to be checked.')
+        linter_stdout = []
+        linter_stdout.append(self.check_mandatory_patterns())
+        linter_stdout.append(self.check_bad_patterns())
+        linter_stdout.append(self.check_newline_at_eof())
```
```suggestion
        linter_stdout = [
        self.check_mandatory_patterns(),
        self.check_bad_patterns(),
        self.check_newline_at_eof()]
```

Should we do it this way^? (Fine with either!)

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/general_purpose_linter.py:
```diff
@@ -694,203 +683,142 @@ def _check_for_mandatory_pattern_in_file(
-        return all_messages
+        if not self.all_filepaths:
+            concurrent_task_utils.log('There are no files to be checked.')
+        linter_stdout = []
```
task_results*

DubeySandeep commented on scripts/linters/general_purpose_linter_test.py:
```diff
@@ -143,615 +143,558 @@ class HTMLLintTests(test_utils.LinterTestBase):
+        self.assert_failed_messages_count(lint_task_report.all_messages, 1)
 
 
 class JsTsLintTests(test_utils.LinterTestBase):
```
Why this file have JSTSLintTest?

Hudda replied:
Because we have some JSTS lint checks in this file. Similarly, we have CSS, HTML and Python lint checks in general_purpose_linter.

DubeySandeep commented on scripts/linters/general_purpose_linter_test.py:
```diff
@@ -143,615 +143,558 @@ class HTMLLintTests(test_utils.LinterTestBase):
-            self.linter_stdout)
-        self.assert_failed_messages_count(self.linter_stdout, 1)
+            lint_task_report.messages)
+        self.assert_failed_messages_count(lint_task_report.all_messages, 1)
```
As mentioned above, please remove these `assert_same_list_elements` & `assert_failed_messages_count` and assert the complete message and other values of the taskResult*

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/general_purpose_linter_test.py:
```diff
@@ -143,615 +143,558 @@ class HTMLLintTests(test_utils.LinterTestBase):
+            '\'ddescribe\'or \'fdescribe\''], lint_task_report.messages)
+        self.assert_failed_messages_count(lint_task_report.all_messages, 1)
 
     def test_invalid_use_of_iit_fit(self):
```
If someone adds new bad_pattern without any test the backend test will fail for coverage? If no, do you have any suggestions on how we can enforce having tests for such cases?

Hudda replied:
No, backend tests won't fail if anyone add a new bad pattern, we have to keep track of it in the code review. 

Sorry, I do not know how to enforce having tests for such cases.

DubeySandeep commented on scripts/linters/html_linter.py:
```diff
@@ -386,105 +343,77 @@ def _get_trimmed_error_output(html_lint_output):
+                [int(s) for s in linter_stdout.split() if s.isdigit()][-2])
+            if error_count:
+                failed = True
+                error_summary.append(error_count)
```
Do we ned `error_summary`?

Hudda replied:
removed.

DubeySandeep reviewed (changes requested):
@Hudda, This looks great, thanks a lot for the work! 

I've reviewed till `scripts/linters/html_linter_test.py` file and left general comments can you please resolve them and follow them for other files. I think we will be able to merge this PR after one more pass!

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Addressed review comments

Hudda commented:
@DubeySandeep I did not make changes to js_linter.py and js_ts_linter_test.py file because the checks of this file are going to be migrated to custom eslint checks so I don't think it is a good idea to spend time on this file

Hudda commented:
@DubeySandeep PTAL!

Hudda commented:
I've applied changes to the js_ts_linter_file.py and does not make private methods to public method. Because checks present in js_ts_linter will be migrated to custom eslint checks and then they will be removed automatically. This task will be done by the first week of October. We have some pros and cons to this approach.

Pros:
1. It will save time and I can work on another PR instead of spending time in this PR.

Cons:
1. JsTs linter test file will take around 2 mins to runs until all checks migrate to custom lint checks.

@bansalnitish is it fine with you?

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,13 +44,52 @@ def log(message, show_time=False):
+                failed.
+            trimmed_messages: list(str). List of error messages that are
+                trimmed to keep main part of messages.
+            full_messages: str|list(str). List of full messages returned by the
```
full_messages should be of one type, change it in the backend test to return a list or change it in linters to return a string* (I would prefer the first one, as the name of the variable is plural.)

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,13 +44,52 @@ def log(message, show_time=False):
+        self.full_messages = full_messages
+
+    @property
+    def all_messages(self):
```
change this to get_report (or something around that) [And it should not be a property!]
also, don't change the full_message property here!

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,13 +44,52 @@ def log(message, show_time=False):
         super(TaskThread, self).__init__()
         self.func = func
-        self.output = None
+        self.task_results = None
```
The default value should be empty list*

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/css_linter_test.py:
```diff
@@ -93,35 +93,30 @@ def mock_communicate():
+        self.assertFalse(lint_task_report[0].failed)
 
     def test_perform_all_lint_checks_with_valid_file(self):
         third_party_linter = css_linter.ThirdPartyCSSLintChecksManager(
```
Instead of running the third_party checks, we should much subprocces and assert the subprocces was called with a given command. [The moto of this file is to have  unit test for the linter script.]

This needs a bit of audit, I would suggest creating an issue and leave it as later work!

Hudda replied:
Done.

DubeySandeep commented on scripts/linters/html_linter.py:
```diff
@@ -84,37 +83,33 @@ def handle_starttag(self, tag, attrs):
         if self.debug:
-            python_utils.PRINT('DEBUG MODE: Start tag_stack')
-            python_utils.PRINT(self.tag_stack)
+            concurrent_task_utils.log('DEBUG MODE: Start tag_stack')
```
Do we really need this message? [Considering this can come between any different logs and it can be messy!]

Can you please create an issue to check whether this is needed and remove it if not required. [Make sure to add a `TODO(#<issue-id>): ....` message on line 109]

Hudda replied:
Done.

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,13 +44,52 @@ def log(message, show_time=False):
+            task.
+        """
+        if self.name:
+            failed_message = (
```
status_message

Hudda replied:
Done.

DubeySandeep reviewed (commented):
Hi @Hudda, This looks good in general, I've left a few minor comments PTAL!

Let's merge this PR once you resolve these last few comments. There are other tasks that need to be done, I've listed a few of them below can you please create an issue for each of them:

1. Change the linter running functionality such that lint function should run in parallel instead of linter class.
2. Remove `assert_same_list_elements` and test the full log of linter script.
3. Add colors for success/failure reports for the lint reports.
4. Change the linter test file such that it doesn't run the third_party linter process while testing!

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -58,14 +99,28 @@ def __init__(self, func, verbose, semaphore, name):
+                for task_result in self.task_results:
+                    # The following section will print the output of the lint
+                    # checks.
+                    if task_result.name:
```
Instead of relying on this field to decide how to print the message, create another filed report_enabled (with True default value of keyword param) and disable it for the backed tests. (Let me know if you need more info on this!)

Hudda replied:
Done.

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Addressed review comments

Hudda committed:
Fixed backend tests

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Added todo comment

Hudda commented:
Created the issue https://github.com/oppia/oppia/issues/10482

Hudda commented:
@DubeySandeep PTAL!

DubeySandeep commented on scripts/concurrent_task_utils.py:
```diff
@@ -42,13 +44,55 @@ def log(message, show_time=False):
+        self.failed = failed
+        self.trimmed_messages = trimmed_messages
+        self.messages = messages
+        self.report_enabled = report_enabled
```
This should be a property of the `TaskThread` object instead of `TaskResult`.

Hudda replied:
@DubeySandeep If I do this then it would not be possible to add the status message in get_report function because this boolean help to decide if it has to add pass/fail status or not

DubeySandeep replied:
If the `report_enabled` is true for taskThread object then log using `get_report` else log the `self.messages`, is that helpful?

Hudda replied:
Done.

DubeySandeep reviewed (approved):
LGTM, just one minor comment, fix that and get this PR merged!

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log

Hudda committed:
Addressed review comment

Hudda committed:
Fixed backend tests

Hudda committed:
Merge remote-tracking branch 'upstream/develop' into print-log
